src/util.jl

```diff
@@ -162,49 +162,101 @@ function solve_triangular_system(d::OrderedDict)
 solve_dependency(system)
 ```
+
+Optionally, one can add specify which subset of variables to solve for.
+Unrequired variables will be ommited in the solution.
+
 """
-function solve_dependency(dd::Dict{T,Set{T}}) where T # is hashable
-    solved = T[]
-    deps = deepcopy(dd)
-    p = length(deps)
-    it = 0
-    while length(deps) > 0 && it <= p
-        it += 1
-        for (k, dep) in deps
-            if length(dep) == 0
-                push!(solved, k)
-                pop!(deps, k)
-                for (l, ldeps) in deps
-                    if k in ldeps
-                        pop!(ldeps, k)
-                    end
+function solve_dependencies(deps::Associative{T,Set{T}}, needed=nothing) where T
+    solution = []
+    if needed == nothing
+        needed = Set(keys(deps))
+    else
+        needed = Set(needed)
+    end
+    while length(needed)>0
+        tt0 = (length(needed), length(solution))
+        if length(needed)==0
+            return solution
+        else
+            for k in needed
+                # check whether k is solved
+                if k in solution
+                    pop!(needed, k)
+                elseif issubset(deps[k], solution)
+                    push!(solution, k)
+                else
+                    needed = union(deps[k], needed)
                 end
             end
+            tt = (length(needed), length(solution))
+            if tt == tt0
+                mis = join([string(e) for e in needed], ", ")
+                exc = TriangularSystemException(mis)
+                throw(exc)
+            end
         end
     end
-    if it >= p+1
-        error("Non triangular system")
+    return solution
+end
+
+list_symbols(expr::Symbol) = list_symbols(:($expr*2))
+
+function get_dependencies(defs::Associative{T,U}) where T where U
+    deps = OrderedDict{Any,Set{Any}}()
+    for (k,v) in (defs)
+        ii = intersect( Set(union( collect( values( Dolang.list_symbols(v) ))... )), Set(keys(defs)))
+        ij = Set(ii)
+        deps[k] = ij
     end
-    return solved
+    deps
+end
+
+function reorder_triangular_block(defs::Associative{T,U}) where T where U
+    deps = get_dependencies(defs)
+    sol = Dolang.solve_dependencies(deps)
+    return OrderedDict((k,defs[k]) for k in sol)
 end
 
-function solve_dependency(dd::Dict{T,Vector{T}}) where T
-    solve_dependency(Dict((k, Set(v)) for (k, v) in dd))
+"""Solves definitions blocks
+
+Keys are timed variables in canonical form (e.g. `(:v,0)`) at date t=0.
+Values are expressions, possibly referencing key variables at different dates.
+The system is recursively solved for the unknowns, by default the keys.
+"""
+function solve_definitions(defs::Associative{Tuple{Symbol, Int}, <:SymExpr}, unknowns=keys(defs))
+    # defs should map timed-vars to expressions.
+    defs = deepcopy(defs)
+    for (v,t) in collect(keys(defs))
+        # t should always be 0
+        @assert t==0
+        for shift in (-1,1)
```

Comment:
Again, is there a reason we restrict shifts to (-1, 1)?

Reply:
Same as before. I'd like to rethink the whole think, ideally in favour of more general pattern-based matching, but for now. +1/-1 doesn't really have disadvantages.